File: 1457-PseudoPalindromicPathsinaBinaryTree/1457-PseudoPalindromicPathsinaBinaryTree.java
/*

1457. Pseudo-Palindromic Paths in a Binary Tree
Medium

Given a binary tree where node values are digits from 1 to 9. A path in the binary tree is said to be pseudo-palindromic if at least one permutation of the node values in the path is a palindrome.

Return the number of pseudo-palindromic paths going from the root node to leaf nodes.

Example 1:

Input: root = [2,3,1,3,1,null,1]
Output: 2 
Explanation: The figure above represents the given binary tree. There are three paths going from the root node to leaf nodes: the red path [2,3,3], the green path [2,1,1], and the path [2,3,1]. Among these paths only red path and green path are pseudo-palindromic paths since the red path [2,3,3] can be rearranged in [3,2,3] (palindrome) and the green path [2,1,1] can be rearranged in [1,2,1] (palindrome).

Example 2:

Input: root = [2,1,1,1,3,null,null,null,null,null,1]
Output: 1 
Explanation: The figure above represents the given binary tree. There are three paths going from the root node to leaf nodes: the green path [2,1,1], the path [2,1,3,1], and the path [2,1]. Among these paths only the green path is pseudo-palindromic since [2,1,1] can be rearranged in [1,2,1] (palindrome).

Example 3:

Input: root = [9]
Output: 1

Constraints:

    The given binary tree will have between 1 and 10^5 nodes.
    Node values are digits from 1 to 9.

*/

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    int res = 0;
    public int pseudoPalindromicPaths (TreeNode root) {
        if(root == null) return res;
        dfs(root, new HashMap<>());
        return res;
    }
    private void dfs(TreeNode node, Map<Integer, Integer> map){
        if(node == null) return;
        map.put(node.val, map.containsKey(node.val) ? map.get(node.val) + 1 : 1);
        if(node.left == null && node.right == null){
            if(isPalindrome(map))
                res++;
        }
        dfs(node.left, map);
        dfs(node.right, map);
        map.put(node.val, map.get(node.val) - 1);
    }
    private boolean isPalindrome(Map<Integer, Integer> map){
        int count = 0;
        for(Map.Entry<Integer, Integer> entry : map.entrySet()){
            if(entry.getValue() % 2 == 0) continue;
            else count++;
            if(count == 2)
               return false;
        }
        return true;
    }
}

File: 1457-PseudoPalindromicPathsinaBinaryTree/1457-PseudoPalindromicPathsinaBinaryTree.py
"""

1457. Pseudo-Palindromic Paths in a Binary Tree

Given a binary tree where node values are digits from 1 to 9. A path in the binary tree is said to be pseudo-palindromic if at least one permutation of the node values in the path is a palindrome.

Return the number of pseudo-palindromic paths going from the root node to leaf nodes.

Example 1:

Input: root = [2,3,1,3,1,null,1]
Output: 2 
Explanation: The figure above represents the given binary tree. There are three paths going from the root node to leaf nodes: the red path [2,3,3], the green path [2,1,1], and the path [2,3,1]. Among these paths only red path and green path are pseudo-palindromic paths since the red path [2,3,3] can be rearranged in [3,2,3] (palindrome) and the green path [2,1,1] can be rearranged in [1,2,1] (palindrome).

Example 2:

Input: root = [2,1,1,1,3,null,null,null,null,null,1]
Output: 1 
Explanation: The figure above represents the given binary tree. There are three paths going from the root node to leaf nodes: the green path [2,1,1], the path [2,1,3,1], and the path [2,1]. Among these paths only the green path is pseudo-palindromic since [2,1,1] can be rearranged in [1,2,1] (palindrome).

Example 3:

Input: root = [9]
Output: 1

Constraints:

    The number of nodes in the tree is in the range [1, 105].
    1 <= Node.val <= 9

"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def pseudoPalindromicPaths (self, root: Optional[TreeNode]) -> int:
        res = [0]
        def psuedoPalin(curr):
            mapp = Counter(curr)
            odd = 0
            for k, v in mapp.items():
                if v % 2 != 0: odd += 1
            return odd <= 1
        def dfs(node, curr):
            if not node: return 
            curr.append(node.val)
            if not node.left and not node.right:
                if psuedoPalin(curr):
                    res[0] += 1
            dfs(node.left, curr)
            dfs(node.right, curr)
            curr.pop()
        dfs(root, [])
        return res[0]
